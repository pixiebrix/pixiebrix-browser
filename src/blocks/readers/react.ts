/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
