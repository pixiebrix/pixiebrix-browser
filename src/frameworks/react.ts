/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import fromPairs from "lodash/fromPairs";

interface DOMFiber {
  _instance: any;
  _currentElement: {
    _owner: DOMFiber;
  };
}

interface ComponentFiber {
  return: ComponentFiber;
  memoizedProps: { [key: string]: unknown };
}

export class ComponentNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ComponentNotFoundError";
  }
}

export function readReactProps(
  fiber: ComponentFiber,
  rootProp?: string
): Record<string, unknown> {
  // getComponentFiber is already returning the "return" for the component
  const memoizedProps = fiber.memoizedProps;
  console.debug("Memoized props", memoizedProps);
  const props = rootProp ? memoizedProps[rootProp] : memoizedProps;
  return fromPairs(Object.entries(props).filter(([key]) => key !== "children"));
}

export function findReactComponent(
  dom: HTMLElement,
  traverseUp = 0
): ComponentFiber {
  // https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging
  // https://stackoverflow.com/a/39165137/402560

  const key = Object.keys(dom).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );
  // @ts-ignore: works in practice
  const domFiber: DOMFiber | null = dom[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found at selector");
  }

  // react <16
  // TODO: find a react 15.0 app to test this on
  if (domFiber._currentElement) {
    throw new Error("Support for React <16 not implemented");
    // let componentFiber = domFiber._currentElement._owner;
    // for (let i = 0; i < traverseUp; i++) {
    //     componentFiber = componentFiber._currentElement._owner;
    // }
    // return componentFiber._instance;
  }

  // react 16+
  const getComponentFiber = (fiber: any) => {
    //return fiber._debugOwner; // this also works, but is __DEV__ only
    let parentFiber = fiber.return;
    while (typeof parentFiber.type == "string") {
      parentFiber = parentFiber.return;
    }
    return parentFiber;
  };
  let componentFiber = getComponentFiber(domFiber);
  for (let i = 0; i < traverseUp; i++) {
    componentFiber = getComponentFiber(componentFiber);
  }
  return componentFiber;
}
