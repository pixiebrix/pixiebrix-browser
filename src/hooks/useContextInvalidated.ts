/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { useState } from "react";
import useAsyncEffect from "use-async-effect";
import {
  onContextInvalidated,
  wasContextInvalidated,
} from "@/errors/contextInvalidated";

export default function useContextInvalidated(): boolean {
  const [invalidated, setInvalidated] = useState(wasContextInvalidated());
  useAsyncEffect(async (isMounted) => {
    await onContextInvalidated();

    if (isMounted()) {
      setInvalidated(true);
    }
  });

  return invalidated;
}
