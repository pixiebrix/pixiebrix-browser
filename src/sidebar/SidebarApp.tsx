/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import store, { persistor } from "@/sidebar/store";
import { Provider } from "react-redux";
import Loader from "@/components/Loader";
import { PersistGate } from "redux-persist/integration/react";
import ConnectedSidebar from "./ConnectedSidebar";
import Header from "./Header";
import ErrorBanner from "./ErrorBanner";
import { MemoryRouter } from "react-router";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import ReduxPersistenceContext, {
  type ReduxPersistenceContextType,
} from "@/store/ReduxPersistenceContext";

registerDefaultWidgets();

const authPersistenceContext: ReduxPersistenceContextType = {
  async flush() {
    await persistor.flush();
  },
};

// Include MemoryRouter because some of our authentication-gate hooks use useLocation. However, there's currently no
// navigation in the SidebarApp
const SidebarApp: React.FunctionComponent = () => (
  <Provider store={store}>
    <PersistGate loading={<Loader />} persistor={persistor}>
      <ReduxPersistenceContext.Provider value={authPersistenceContext}>
        <MemoryRouter>
          <ErrorBanner />
          <Header />
          <ConnectedSidebar />
        </MemoryRouter>
      </ReduxPersistenceContext.Provider>
    </PersistGate>
  </Provider>
);

export default SidebarApp;
