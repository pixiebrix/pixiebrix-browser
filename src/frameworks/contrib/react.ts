/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { pickBy } from "lodash";
import { ComponentNotFoundError, ignoreNotFound } from "@/frameworks/errors";
import { RootInstanceVisitor } from "@/frameworks/scanner";
import { ReadableComponentAdapter, traverse } from "@/frameworks/component";
import { isNode } from "@/frameworks/dom";

// React architecture references:
// https://github.com/acdlite/react-fiber-architecture
// https://indepth.dev/posts/1008/inside-fiber-in-depth-overview-of-the-new-reconciliation-algorithm-in-react#memoizedprops
// https://github.com/Venryx/mobx-devtools-advanced/blob/master/Docs/TreeTraversal.md
// https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging

interface RootInstance {
  _reactRootContainer: unknown;
}

/**
 * Structure from:
 * - https://github.com/facebook/react/blob/50d9451f320a9aaf94304209193562cc385567d8/packages/react-devtools-shared/src/backend/legacy/renderer.js#L739
 * - https://github.com/facebook/react/blob/v15.0.0/src/renderers/shared/reconciler/ReactReconciler.js
 */
interface LegacyInstance {
  /**
   * Public instance
   */
  _instance: unknown;

  _currentElement: {
    _owner: LegacyInstance;
  };
}

/**
 * Type definition from: https://github.com/facebook/react/blob/4c6470cb3b821f3664955290cd4c4c7ac0de733a/packages/react-reconciler/src/ReactInternalTypes.js#L62
 */
interface Fiber {
  /**
   * The Fiber to return to after finishing processing this one.
   * This is effectively the parent, but there can be multiple parents (two)
   * so this is only the parent of the thing we're currently processing.
   * It is conceptually the same as the return address of a stack frame.
   */
  return: Fiber | null;

  /**
   * The props used to create the output.
   */
  memoizedProps: { [prop: string]: unknown };

  /**
   * The resolved function/class associated with this fiber.
   */
  type: string | object;

  /**
   * The local state associated with this fiber.
   */
  stateNode: Node | object;
}

export function isManaged(node: Node): boolean {
  return Object.keys(node).some((key) =>
    key.startsWith("__reactInternalInstance$")
  );
}

export function hasReactProps(fiber: Fiber): boolean {
  return Object.keys(fiber.memoizedProps).some((x) => x !== "children");
}

export function readReactProps(fiber: Fiber): { [prop: string]: unknown } {
  return pickBy(fiber.memoizedProps, (value, key) => key !== "children");
}

function getComponentFiber(fiber: Fiber): Fiber {
  // return fiber._debugOwner; // this also works, but is __DEV__ only
  let parentFiber = fiber.return;
  while (typeof parentFiber.type === "string") {
    // string for HTML nodes, so traverse
    parentFiber = parentFiber.return;
  }
  return parentFiber;
}

export function findReactComponent(node: Node, traverseUp = 0): Fiber {
  // https://stackoverflow.com/a/39165137/402560
  const key = Object.keys(node).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );

  const domFiber: Fiber | LegacyInstance | null = (node as any)[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found for element");
  }

  if ("_currentElement" in domFiber) {
    // FIXME: test this for React <16
    const owner = (x: LegacyInstance) => x._currentElement._owner;
    const fiber = traverse(owner, owner(domFiber), traverseUp);
    return fiber._instance as Fiber;
  } else {
    return traverse(getComponentFiber, getComponentFiber(domFiber), traverseUp);
  }
}

export class ReactRootVisitor implements RootInstanceVisitor<RootInstance> {
  public rootInstances: RootInstance[] = [];
  visit(node: Element | Node): boolean {
    if ("_reactRootContainer" in node) {
      this.rootInstances.push(node);
      return false;
    }
    return true;
  }
}

export const adapter: ReadableComponentAdapter<Fiber> = {
  isManaged,
  getComponent: (node) => ignoreNotFound(() => findReactComponent(node, 0)),
  getParent: getComponentFiber,
  getNode: (instance) =>
    isNode(instance.stateNode) ? instance.stateNode : null,
  getData: readReactProps,
  hasData: hasReactProps,
};

export default adapter;
