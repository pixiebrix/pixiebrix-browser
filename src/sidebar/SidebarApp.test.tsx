/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import SidebarApp from "@/sidebar/SidebarApp";
import { render } from "@testing-library/react";
import useContextInvalidated from "@/hooks/useContextInvalidated";

jest.mock("@/store/optionsStore", () => ({
  persistor: {
    flush: jest.fn(),
  },
}));

jest.mock("@/hooks/useContextInvalidated", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("SidebarApp", () => {
  test("it renders", () => {
    const rendered = render(<SidebarApp />);
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("it renders error when context is invalidated", () => {
    (useContextInvalidated as jest.Mock).mockReturnValue(true);
    const rendered = render(<SidebarApp />);
    expect(rendered.asFragment()).toMatchSnapshot();
  });
});
