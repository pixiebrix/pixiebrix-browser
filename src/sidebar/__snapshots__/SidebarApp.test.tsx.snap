// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarApp renders 1`] = `
<DocumentFragment>
  <div
    class="full-height"
  >
    <div
      class="d-flex p-2 justify-content-between align-content-center"
    >
      <button
        class="button btn btn-link btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-double-right fa-w-14 fa-lg"
          data-icon="angle-double-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="align-self-center"
      >
        <img
          alt="PixieBrix logo"
          class="logo"
          src="test-file-stub"
        />
      </div>
      <a
        class="button btn btn-link btn-sm"
        href="/options.html"
        target="_blank"
      >
        <span>
          Options 
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-cog fa-w-16 "
            data-icon="cog"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
              fill="currentColor"
            />
          </svg>
        </span>
      </a>
    </div>
    <div
      class="full-height"
    >
      <div>
        <div
          class="container"
        >
          <div
            class="mt-4 row"
          >
            <div
              class="col"
            >
              <h4
                class="display-6"
              >
                Activate an Official Blueprint
              </h4>
              <p>
                <span
                  class="text-primary"
                >
                  The easiest way to start using PixieBrix!
                </span>
                 Activate a pre-made blueprint from the Public Marketplace.
              </p>
              <a
                class="btn btn-info"
                href="https://www.pixiebrix.com/marketplace/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Open Marketplace 
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-external-link-alt fa-w-16 "
                  data-icon="external-link-alt"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                    fill="currentColor"
                  />
                </svg>
              </a>
            </div>
          </div>
          <div
            class="mt-4 row"
          >
            <div
              class="col"
            >
              <h4
                class="display-6"
              >
                Create your Own
              </h4>
              <p>
                Follow the Quick Start Guide to start creating your own bricks in minutes.
              </p>
              <a
                class="btn btn-info"
                href="https://docs.pixiebrix.com/quick-start-guide"
                rel="noopener noreferrer"
                target="_blank"
              >
                Open Quick Start Guide 
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-external-link-alt fa-w-16 "
                  data-icon="external-link-alt"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                    fill="currentColor"
                  />
                </svg>
              </a>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="text-center col"
            >
              <img
                alt="Marketplace"
                src="test-file-stub"
                width="300"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
